deepcopy

hydroflame opened the pull request:
added deep copy

hydroflame commented:
aahh, there might be a problem with string copy, if the original memory was RW, the copy isn't

hydroflame commented:
I don't need 3 review, just one would be fine

ScottMansfield commented on deepcopy.go:
```diff
@@ -0,0 +1,174 @@
+		return s.Interface()
+
+	case reflect.Chan:
+		// verify if we already have copied that string
```
somewhat confusing comments, i assume it was just a copy/paste error.

ScottMansfield commented on deepcopy.go:
```diff
@@ -0,0 +1,174 @@
+		return ch.Interface()
+
+	case reflect.Map:
+		// verify if we already have copied that string
```
comment

ScottMansfield commented on deepcopy.go:
```diff
@@ -0,0 +1,174 @@
+		return string(src)
+
+	case reflect.Func:
+		// func cannot be copied per say, that means closures will remain closure.
```
this is extremely nitpick-y but it's "per se" not "per say"

ScottMansfield commented on deepcopy.go:
```diff
@@ -0,0 +1,174 @@
+		}
+		return s.Interface()
+	case reflect.Ptr:
+		// verify if we already have copied that string
```
comment

ScottMansfield commented on deepcopy.go:
```diff
@@ -0,0 +1,174 @@
+	case reflect.Complex128:
+		return complex128(value.Complex())
+
+	case reflect.Uintptr:
```
In the documentation for the function you mention it doesn't copy these, I'd like to see some comments saying so here as well.

ScottMansfield commented on deepcopy.go:
```diff
@@ -0,0 +1,174 @@
+
+		return interf
+	case reflect.Interface:
+		panic("Is that even reacheable ?")
```
This ought to be more descriptive so if it does get reached it's not a mystery what triggered it.

ScottMansfield commented on deepcopy_test.go:
```diff
@@ -0,0 +1,150 @@
+)
+
+func TestDeepCopy(t *testing.T) {
+	tests := []struct {
```
I don't see a test verifying the unsafe.Pointer is *not* copied

ScottMansfield commented on deepcopy_test.go:
```diff
@@ -0,0 +1,150 @@
+	}
+}
+
+// TestDeepCopyChannel tests the copying of channels, reflect.DeepEqual doesn't
```
You could add a test with a buffered channel, too, to check if the contents are copied over.

ScottMansfield commented on deepcopy_test.go:
```diff
@@ -0,0 +1,150 @@
+
+// TestDeepCopyRing tests that a circular reference doesn't go out of control
+// and allocate everything.
+func TestDeepCopyRing(t *testing.T) {
```
this is a great idea for a test.

hydroflame commented:
@ScottMansfield I've addressed most of your comments and good catch, there was a bug with unsafe-like pointers. However I have not fixed the copying of channel content as it's kindof a mystery as to how this would even be done. channel are intrinsics to the compiler and it would be pretty unstable to try to copy the content. I have however explicitly spelled that out in the comment for the function
